src/main/java/com/atlassian/bitbucket/jenkins/internal/client/ModernBitbucketBuildStatusClientImpl.java

```diff
@@ -1,26 +1,61 @@
 package com.atlassian.bitbucket.jenkins.internal.client;
 
+import com.atlassian.bitbucket.jenkins.internal.client.exception.BitbucketClientException;
 import com.atlassian.bitbucket.jenkins.internal.model.BitbucketBuildStatus;
+import com.atlassian.bitbucket.jenkins.internal.provider.DefaultInstanceIdentityProvider;
+import com.atlassian.bitbucket.jenkins.internal.provider.InstanceIdentityProvider;
+import com.google.common.annotations.VisibleForTesting;
 import okhttp3.HttpUrl;
 
+import java.security.Signature;
+import java.security.interfaces.RSAPrivateKey;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+
 import static java.util.Objects.requireNonNull;
 import static org.apache.commons.lang3.StringUtils.stripToNull;
 
 public class ModernBitbucketBuildStatusClientImpl implements BitbucketBuildStatusClient {
 
+    private static final String BBS_BUILD_STATUS_SIGNATURE_ID = "BBS-Signature";
     private static final String BUILD_STATUS_VERSION = "1.0";
     private final BitbucketRequestExecutor bitbucketRequestExecutor;
-    private final String revisionSha;
+    private final InstanceIdentityProvider instanceIdentityProvider;
     private final String projectKey;
     private final String repoSlug;
+    private final String revisionSha;
 
-    ModernBitbucketBuildStatusClientImpl(BitbucketRequestExecutor bitbucketRequestExecutor, String projectKey, String repoSlug, String revisionSha) {
+    @VisibleForTesting
+    ModernBitbucketBuildStatusClientImpl(BitbucketRequestExecutor bitbucketRequestExecutor, String projectKey,
+                                         String repoSlug, String revisionSha,
+                                         InstanceIdentityProvider instanceIdentityProvider) {
         this.bitbucketRequestExecutor = requireNonNull(bitbucketRequestExecutor, "bitbucketRequestExecutor");
+        this.instanceIdentityProvider = requireNonNull(instanceIdentityProvider, "instanceIdentityProvider");
         this.revisionSha = requireNonNull(stripToNull(revisionSha), "revisionSha");
         this.projectKey = requireNonNull(stripToNull(projectKey), "projectKey");
         this.repoSlug = requireNonNull(stripToNull(repoSlug), "repoSlug");
     }
 
+    ModernBitbucketBuildStatusClientImpl(BitbucketRequestExecutor bitbucketRequestExecutor, String projectKey,
+                                         String repoSlug, String revisionSha) {
+        this(bitbucketRequestExecutor, projectKey, repoSlug, revisionSha, new DefaultInstanceIdentityProvider());
+    }
+
+    public Map<String, String> generateSignature(BitbucketBuildStatus buildStatus) {
+        Map<String, String> headers = new HashMap<>();
+        try {
+            RSAPrivateKey key = instanceIdentityProvider.getInstanceIdentity().getPrivate();
+            Signature sig = Signature.getInstance("SHA256with" + key.getAlgorithm());
+            sig.initSign(key);
+            sig.update(buildStatus.getSignature().getBytes("UTF-8"));
```

Comment:
nit: non-string charset encoding:
```suggestion
            sig.update(buildStatus.getSignature().getBytes(UTF_8));
```

Reply:
As a side-effect, this removes one of the checked exceptions from your catch clause: `UnsupportedEncodingException`